toolkit/tools/imagegen/diskutils/initramfs.go

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Utility to encrypt disks and partitions
+
+package diskutils
+
+import (
+	"bytes"
+	"io"
+	"os"
+
+	"github.com/cavaliercoder/go-cpio"
+	"github.com/klauspost/pgzip"
+	"microsoft.com/pkggen/internal/logger"
+)
+
+// InitramfsMount represented an editable initramfs
+type InitramfsMount struct {
+	pgzWriter           *pgzip.Writer
+	cpioWriter          *cpio.Writer
+	outputBuffer        *bytes.Buffer
+	initramfsOutputFile *os.File
+}
+
+// CreateInitramfs creates a new initramfs
+func CreateInitramfs(initramfsPath string) (initramfs InitramfsMount, err error) {
+	// Initramfs traditionally is -rw-------
+	const initramfsModeBits = os.FileMode(0600)
+	// Caller must use InitramfsMount.Close() to clean up the outputs
+	initramfs.outputBuffer = new(bytes.Buffer)
+	initramfs.pgzWriter = pgzip.NewWriter(initramfs.outputBuffer)
+	initramfs.cpioWriter = cpio.NewWriter(initramfs.pgzWriter)
+
+	initramfs.initramfsOutputFile, err = os.OpenFile(initramfsPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, initramfsModeBits)
+
+	return
+}
+
+// OpenInitramfs makes an existing initramfs editable
+func OpenInitramfs(initramfsPath string) (initramfs InitramfsMount, err error) {
+	logger.Log.Debugf("Opening intramfs '%s'", initramfsPath)
+	inputFile, err := os.Open(initramfsPath)
+	if err != nil {
+		return
+	}
+	defer inputFile.Close()
+
+	gzReader, err := pgzip.NewReader(inputFile)
+	if err != nil {
+		return
+	}
+	defer gzReader.Close()
+	cpioReader := cpio.NewReader(gzReader)
+	//cpio.Reader has no Close() function to defer
+
+	// Caller must use InitramfsMount.Close() to clean up the outputs
+	initramfs.outputBuffer = new(bytes.Buffer)
+	initramfs.pgzWriter = pgzip.NewWriter(initramfs.outputBuffer)
+	initramfs.cpioWriter = cpio.NewWriter(initramfs.pgzWriter)
+
+	var bytesIO int64
+	var nextFileHeader *cpio.Header
+	for {
+		var linkPayload []byte
+
+		nextFileHeader, err = cpioReader.Next()
+		if err == io.EOF {
+			err = nil
+			break
+		}
```

Comment:
Would it be possible to put that into the `for` line?

Reply:
Handing the error is a bit of a pain unfortunately. I'll give it a think though.

---
In reply to: [550331448](https://github.com/microsoft/CBL-Mariner/pull/481#discussion_r550331448) [](ancestors = 550331448)

Reply:
Going to leave this as is I think, unless you can see a cleaner way to organizing it.

That said, I've added comments to explain the flow and better scoped some variables into the loop.

---
In reply to: [550696501](https://github.com/microsoft/CBL-Mariner/pull/481#discussion_r550696501) [](ancestors = 550696501,550331448)